File: desktop/src/shared/localUIExposure.ts
export type LocalUIExposure = Readonly<{
	scope: 'loopback' | 'network';
	transport: 'plaintext';
	password_required: boolean;
}>;

export function parseLocalUIExposure(value: unknown): LocalUIExposure {
	if (!value || typeof value !== 'object') {
		throw new Error('missing Local UI exposure');
	}
	const record = value as Record<string, unknown>;
	const scope = String(record.scope ?? '').trim();
	const transport = String(record.transport ?? '').trim();
	if (scope !== 'loopback' && scope !== 'network') {
		throw new Error('invalid Local UI exposure scope');
	}
	if (transport !== 'plaintext') {
		throw new Error('invalid Local UI exposure transport');
	}
	if (typeof record.password_required !== 'boolean') {
		throw new Error('invalid Local UI exposure password requirement');
	}
	if (scope === 'network' && record.password_required !== true) {
		throw new Error('network Local UI exposure requires password authentication');
	}
	return {
		scope,
		transport,
		password_required: record.password_required,
	};
}

export function networkLocalUIExposureActive(exposure: LocalUIExposure | null | undefined): boolean {
	return exposure?.scope === 'network' && exposure.transport === 'plaintext';
}

File: internal/runtimemanagement/local_ui_exposure.go
package runtimemanagement

import "fmt"

type LocalUIExposureScope string

const (
	LocalUIExposureScopeLoopback LocalUIExposureScope = "loopback"
	LocalUIExposureScopeNetwork  LocalUIExposureScope = "network"
	LocalUITransportTLS                               = "tls"
	LocalUITransportHTTP                              = "http"
)

// LocalUIExposure is the canonical runtime security posture projected to every
// Local UI status surface.
type LocalUIExposure struct {
	Scope            LocalUIExposureScope `json:"scope"`
	Transport        string               `json:"transport"`
	PasswordRequired bool                 `json:"password_required"`
}

func NewLocalUIExposure(protocol string, network bool, passwordRequired bool) LocalUIExposure {
	scope := LocalUIExposureScopeLoopback
	if network {
		scope = LocalUIExposureScopeNetwork
	}
	transport := protocol
	if protocol == "https" {
		transport = LocalUITransportTLS
	}
	return LocalUIExposure{
		Scope:            scope,
		Transport:        transport,
		PasswordRequired: passwordRequired,
	}
}

func (e LocalUIExposure) Validate() error {
	if e.Scope != LocalUIExposureScopeLoopback && e.Scope != LocalUIExposureScopeNetwork {
		return fmt.Errorf("invalid Local UI exposure scope %q", e.Scope)
	}
	if e.Transport != LocalUITransportTLS && e.Transport != LocalUITransportHTTP {
		return fmt.Errorf("invalid Local UI exposure transport %q", e.Transport)
	}
	if e.Scope == LocalUIExposureScopeNetwork && !e.PasswordRequired {
		return fmt.Errorf("network Local UI exposure requires password authentication")
	}
	return nil
}

func (e LocalUIExposure) IsNetwork() bool {
	return e.Scope == LocalUIExposureScopeNetwork
}
